### ISSUE TITLE
Code for indiscriminate attack

### ISSUE THREAD

**Issue by Cascol-Chen** [inspire-group/tta_risk#2]
It's a really nice work that reminds the adversarial vulnerability of TTA.

I want to ask for the source code of indiscriminate attack. You have provided a reply for this in another [issues](https://github.com/inspire-group/tta_risk/issues/1), which changes the loss function in generate_attacks function to the following.
```python
loss = nn.CrossEntropyLoss(reduction='none')(out[:-source], y[:-source]).clamp(min=0,max=5)
loss = loss.sum()
adv.data = (adv + alpha*adv.grad.detach().sign()).clamp(-epsilon,epsilon)
```

Nevertheless, I find that this loss cannot provide gradients for _adv_ when I pass the original ViT into the generate_attacks function (since _adv_ modify out[-source:], i.e., the latter inputs). Meanwhile, when I pass the Tented model into the generate_attacks function, RuntimeError occurs, "Trying to backward through the graph a second time, but the saved intermediate results have already been freed. Specify retain_graph=True when calling .backward() or autograd.grad() the first time."

Could you provide more explanations on how to conduct indiscriminate attack? Or is the attack only available on BN models?

**Comment by tongwu2020** [comment]
Hi,

Which torch version do you use? Could you try loss.backward(retain_graph=True) ?

Best,
Tong
